### DISCUSSION TITLE
How to append versions to URL?

### DISCUSSION THREAD

**Discussion by PiliNi** [microsoft/typespec#2665] [Q&A] [answered]
I am looking for a decorator to append versions to the URL.
My case is as below:
```
@versioned(Versions)
@route("petstore")
namespace PetStoreService;

enum Versions{
  v1,v2,v3
}

interface Pets extends Resource.ResourceOperations<Pet, Error> {}
```
There will generate three openapi files(v1,v2,v3) after executing `tsp complie .` and these files include the same interfaces below:
[POST] /petstore
[GET] /petstore
[GET] /petstore/{id}
[PATCH] /petstore/{id}
[DELETE] /petstore/{id}

My question is how to automatically append the version enum to the URL like:
```
// in openapi.v1.yaml like
[POST] /petstore/v1
[GET] /petstore/v1
[GET] /petstore/{id}/v1
[PATCH] /petstore/{id}/v1
[DELETE] /petstore/{id}/v1
```
```
// in openapi.v2.yaml like
[POST] /petstore/v2
[GET] /petstore/v2
[GET] /petstore/{id}/v2
[PATCH] /petstore/{id}/v2
[DELETE] /petstore/{id}/v2
...
```
```
// in openapi.v3.yaml like
[POST] /petstore/v3
[GET] /petstore/v3
[GET] /petstore/{id}/v3
[PATCH] /petstore/{id}/v3
[DELETE] /petstore/{id}/v3
...
```

**Comment by markcowl** [comment, marked as answer]
The short answer is that there isn't a way to *automatically* do this yet, however, you can achieve a similar thing by using explict `@route` decoration with versioned interfaces or operations, for example [this playground shows versioning interfaces](https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CtIZcmVzdNUZdmVyc2lvbmluZyI7Cgp1c2luZyBUeXBlU3BlYy5IdHRwO9AVUmVzdNEVVslH1TDEBW91cmNlOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCn0pCkDnAJxlZCjHX3MpCkByb3V0ZSgicGV0c3RvcmUiKQpuYW1lc3BhY2UgUGV0U8QVx0c7CgplbnVtIMg%2BIMRydjEsxAYyxQYzLAp9Cgptb2RlbMQ9xSFAa2V5CiAgxFY6IHN0cuUA1MU5UHJvcM0TQGFkZOsAni52MinFXN8pcy52M8UpM84pfQoKQGVycm9y5wCSRcQM5QCUY29k6wCNICBtZXNzYWfLE8Q7cmVtb3bNZjLqATAvdjEiKQpAcHJvamVjdGVkTmFtZSgianNvbiIsICJQZXRzQ29sbGVjdGlvbiIpCmludGVyZucBVMsbVjEgZXh0ZW5kcyDoAdPKHU9wZXJhxAlzPFBldCzmAMA%2BIHv%2FAJ0oInvkAW19%2FwCjZXRz8QCZVjHmAbNwYXRjaMQJdXBkYXRlc%2BgAnChQZXTkAWrGFyguLukCiVBhcmFtZXRlcuUArD4sIEBib2R5IMQFOsRbKcUGIHzmAMXmAWHzAe7zANgz7AF1Mv8A0v8BdcYbVjL%2FAXX4AXX%2FALHyALHoAYn%2FALfzAYky%2FwGJ%2FwGJ%2FwGJ%2FwGJ7QFzM%2F8AvP8Bc8YbdjP%2FAXP%2FAXP8AV3%2FAKHzAV0z%2FwFd%2FwFd%2FwFd7QFd&e=%40typespec%2Fopenapi3&options=%7B%7D)

Creating templates based on these definitions would make this a bit easier to express.  In the future, it may be possible to use string interpolation to represent the version value in the operation route, which would remove the need to represent the versioned operatiosn separately.
